### ISSUE TITLE
[Request]: Fall back to a different download/install method if current method fails

### ISSUE THREAD

**Issue by vavavr00m** [asheroto/winget-install#45]
### Describe the feature or improvement you're requesting

Freshly installed W11 23h2 and fully updated. It had winget installed at first but somehow it got messed up and just refuses to be reinstalled. Tried your script but it just terminates and checking the winget command after still says winget is not a recognized command. What worked for me was manually downloading the .msixbundle from the official repo, launched it and clicked reinstall, manually add the path environment variable: `%UserProfile%\AppData\Local\Microsoft\WindowsApps`. Maybe add that to the script?

### Additional context

Can't file a bug report now since I didn't know there was troubleshooting guide until a Feature Request was initiated. Just to add, this script was also downloaded manually and ran locally and got the same result.

**Comment by asheroto** [comment]
Gotcha. I will try installing fresh Windows 11 23H2 to try to replicate the result in a virtual machine.

The window terminating is annoying! It's on the to do list to fix, for now you can do  `-wait`.

If you were able to get it installed manually then I should be able to fix the script to make it work.

Thanks for the heads-up, will check later tonight or tomorrow.

**Comment by asheroto** [comment]
FYI, ChrisTitusTech just uses this script. 😊

I downloaded the ISO from here on 2024-06-25:
https://www.microsoft.com/en-us/software-download/windows11

Checksum sha256 is `36de5ecb7a0daa58dce68c03b9465a543ed0f5498aa8ae60ab45fb7c8c4ae402`. Version 23H2 (OS Build 22631.2861).

Installed in a VM using Hyper-V.

winget is already included in later versions of Windows 11, so no need to install it on that. Fresh install, `winget` command works immediately after installing Windows.

![image](https://github.com/asheroto/winget-install/assets/49938263/4e2c144b-d244-4964-8413-0beb652f2987)

Then I ran this command...

```powershell
&([ScriptBlock]::Create((irm asheroto.com/winget))) -Force -Wait
```

I found that `winget.exe` still works after running.

If you have a different ISO version, please let me know the sha256 checksum and I can try and find that version to download. It may be an older version.

**Comment by asheroto** [comment]
Gotcha, thanks for the info!

I created an ISO using the Media Creation Tool just now and installed Windows in a VM.

winget still works. Running the script with the command I used before also works...

![image](https://github.com/asheroto/winget-install/assets/49938263/d02d7cb4-1fbc-4695-8321-372e5a2d326f)

Maybe it's another script that's causing the issue?

Looks like when winget is installed it adds its own path to the user's PATH variable:
![image](https://github.com/asheroto/winget-install/assets/49938263/c95c7542-4bc4-4d4b-bfb3-633aa8188d0f)

That's okay though, I could see where this might occur for some oddball reason, even if we can't find that reason right now. I went ahead and added the ability to check the PATH variable and add it in if it doesn't exist.

Run this for me and tell me if it works now:

```powershell
$url = "https://github.com/asheroto/winget-install/blob/4ac35195876d9b1206283ae572024159b189a793/winget-install.ps1"
&([ScriptBlock]::Create((irm $url))) -Force -Wait
```

**Comment by asheroto** [comment]
Oops, fixed the URL, try the updated one.

**Comment by asheroto** [comment]
Hmm. I don't understand why it works for me and not you. When using the media creation tool, you select Make ISO, 23H2, right?

What language and locale / keyboard? Maybe that's why.

**Comment by asheroto** [comment]
Hmm, that's the same as me.

After installing, is that the first script you run? Seems like something is interfering here.

**Comment by asheroto** [comment]
What's really odd is that since winget is included in Windows 11, the path should absolutely be in the environment variables. So it's almost like they were removed. At first you said it installed but just didn't run. So I'm guessing other scripts modified or interfered with winget.

Since I'm using the same exact image but it works fine, the issue must be caused by some outside script. Not sure if something else corrupted Windows or what's going on.

I did add the ability to confirm the environment path. Apart from that, not much else changed in the script. So this should be good. I'll run a few more tests and then release this update.

Thanks for the heads-up!

**Comment by asheroto** [comment]
Using the paths you mentioned would probably be better. However, changing a username won't result in the file path changing. If Bob changes his username to Fred, the user profile path will still be C:\Bob.

I'll take a closer look at this when I get a chance. Which lines in the code are you referring to?

**Comment by asheroto** [comment]
I think you might be using an old version. Are you using version 5.0.5?

The only path set is if one is not already set by the winget installer. That path does appear to contain the username but is not something that is set by this script.

https://github.com/asheroto/winget-install/blob/68c485b6a4b1bcf137c999f747e913f49238cc25/winget-install.ps1#L1216

**Comment by asheroto** [comment]
> This does not answer the question: what if the username was changed - wouldn't it break winget, thus, would need to re-run the script?

No, changing a user's username does not change the file path. I answered this earlier. 😊 If you change your username from Bob to Fred, the file path of user "Fred" will still be `C:\Users\Bob`. The only way this changes is if you rename the user folder. Many things will break, not just winget. That's why Windows doesn't change the file path itself.

I do agree that the environment variable should not use the username, but my point is that it's not this script that is responsible for the environment variable that you are concerned about. 😊 I do think you have a valid concern, but think you should bring it up with [winget-cli](https://github.com/microsoft/winget-cli) itself. Otherwise we're just chasing problems caused by the winget repair command.

The reason for this is that if even we make the change with winget-installer, someone would still have the same issue when winget was installed manually, or in later Windows editions when winget installed by default.

**Comment by asheroto** [comment]
So if we added this path to the PATH variable, it would prevent the issues, yes?

`%LOCALAPPDATA%\Microsoft\WindowsApps`

Seems simple enough. It wouldn't fix the problem of other folders being wrong, but at least winget would work if someone change their username and folder or if they used non-Latin characters.

**Comment by asheroto** [comment]
@vavavr00m 

Can you try running this version?

https://raw.githubusercontent.com/asheroto/winget-install/e0a4521e9e4f31b79b41c71ebc3963a4860e22ab/winget-install.ps1

**Comment by asheroto** [comment]
Ugh that's too bad. No worries, I think it's good now.

**Comment by asheroto** [comment]
I believe version 5.0.7 fixes this. If not let me know and I will reopen.
